src/coreclr/jit/flowgraph.cpp

```diff
@@ -711,6 +711,79 @@ bool Compiler::fgIsBlockCold(BasicBlock* blk)
     return ((blk->bbFlags & BBF_COLD) != 0);
 }
 
+//------------------------------------------------------------------------
+// fgIsSafeToRemoveCastOnAssignment: Determine whether "tree" is an assignment whose
+//                                   right-hand side is a cast that can be removed.
+//
+// Arguments:
+//    tree - The tree node under consideration
+//
+// Return Value:
+//    Returns true if the cast can be safely removed.
+//
+bool Compiler::fgIsSafeToRemoveCastOnAssignment(GenTree* tree)
+{
+    assert(tree->OperIs(GT_ASG));
+
+    GenTree* const op1 = tree->gtGetOp1();
+    GenTree* const op2 = tree->gtGetOp2();
+
+    GenTree* const effectiveOp1 = op1->gtEffectiveVal();
+
+    if (!effectiveOp1->OperIs(GT_IND, GT_LCL_VAR))
+        return false;
+
+    if (effectiveOp1->OperIs(GT_LCL_VAR) &&
+        !lvaGetDesc(effectiveOp1->AsLclVarCommon()->GetLclNum())->lvNormalizeOnLoad())
+        return false;
+
+    // If we are storing a small type, we might be able to omit a cast.
+    if (!varTypeIsSmall(effectiveOp1))
```

Comment:
Hmm, some of these conditions are redundant.

`varTypeIsSmall(effectiveOp1)` implies `varTypeIsIntegral(effectiveOp1)` (checked below).